File: Assignment_6/MichaelPeterson/Java/Assignment_6/src/Assignment6.java
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.PBEKeySpec;
import java.io.*;
import java.nio.file.*;
import java.nio.file.attribute.BasicFileAttributeView;
import java.nio.file.attribute.FileOwnerAttributeView;
import java.text.SimpleDateFormat;
import java.util.Arrays;
import java.util.Date;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.security.SecureRandom;

public class Assignment6 implements Runnable{

    private static String OS = System.getProperty("os.name").toLowerCase();
    private static String userName = System.getProperty("user.name").toLowerCase();
    private static final int STRING_INPUT_LENGTH = 50;
    private static final int INT_INPUT_LENGTH    = 11;
    private static final int PWD_INPUT_LENGTH    = 56;
    private static boolean TEST                  = false;

    File logFile;


    public static void main(String[] args) {
        (new Assignment6()).run();
    }


    private class RegexValidator{
        public String regex;
        public Boolean shouldMatch;
        public String failDescription;

        public RegexValidator(String inRegex, Boolean inShouldMatch, String inFailDescription){
            if(inRegex == null || inShouldMatch == null || inFailDescription == null)
                throw new NullPointerException();

            regex = inRegex;
            shouldMatch = inShouldMatch;
            failDescription = inFailDescription;
        }

        public Boolean validate(String input){
            Matcher matcher = Pattern.compile(regex).matcher(input);

            return shouldMatch == matcher.matches();
        }
    }

    private enum FileType{
        Input,
        Output
    }

    private enum UserInputStatus{
        Empty,
        WithinLimits,
        TooLarge
    }

    private class UserInputResult{
        String value;
        UserInputStatus status;

        public UserInputResult(String inValue, UserInputStatus inStatus){
            value = inValue;
            status = inStatus;
        }
    }

    @Override
    public void run() {
        // First things first, make sure this isn't being ran as root
        if("root".equals(userName)){
            System.err.println("This program must not be run as the root user!");
            System.exit(1);
        }

        if (!(OS.indexOf("nix") >= 0 || OS.indexOf("nux") >= 0 || OS.indexOf("aix") > 0)) {
            System.err.println("This software is designed to be run on a Linux operating system.");
            System.exit(1);
        }

        String firstName, lastName;

        // Files
        File inputFile, outputFile;

        logFile = new File("error.log");
        if(!logFile.exists()) {
            try {
                logFile.createNewFile();
            }
            catch(Exception e){
                System.err.println(e.getMessage());
                System.exit(1);
            }
        }

        if(!logFile.canWrite()){
            System.err.println("Unable to write to log file \"error.log\"");
            System.exit(1);
        }


        // regexes
        String nameRegex = "^[a-zA-Z]{1,}$";
        String numberRegex = "^(()|\\+|\\-)[0-9]{1,10}$";
        String filenameRegex_InCurrentDir = "^(\\.\\/|[^\\/])[a-zA-Z0-9\\s]+(\\.[a-zA-Z]{1,4})$";
        String filenameRegex_HasAcceptedExtension = "[^\\.]{1,}\\.(text|txt)$"; // White list of valid extensions, we need more ideas ...
        String filenameRegex_HasRelativePath = "\\.\\.";
        String passwordRegex = "^(?=(.*[a-z].*))(?=(.*[A-Z].*))(?=.*\\d.*)(?=.*\\W.*)[a-zA-Z0-9\\S]{12,56}$";

        if(TEST){
            String nameTestCases[]     = {"", "\0", "\n", "Jhon", " Jhon", "1ll124lk1 1 4l ", "bob", "a", "13", "bob ", "0", "..", "Jo hn"};
            String numberTestCases[]   = {"", "\0", "\n", "0", " 0", "0 ", " 0 ", "-0", "+0", "123", "-123", "+123", "- 123", "123-", "+0123546789", "+01235467891", " ", "apple", "abc", "-12345678901", "--12345678", "-+34567890", "- 34567890", "123456789-","123456789 +", "123456789++", "-123456789-", "(123*456)", "0000000000", "-0000000000", "*1234567890", "1-23456789", "--", "--0", "0000000000000001"};
            String filenameTestCases[] = {"", "\0", "\n", "file.dog", "file.Txt", "file.txt", "./file/../.txt", "/file.txt", "./file.txt", "../file.txt", "/../file.txt", "./ /", "\\/bin\\/file.txt", "*.txt", "file.*"};
            String passwordTestCases[] = {"", "\0", "\n", "abcde", "Slg3k4k23j4Dkj4k23j4kn234jh", "2l23k23ISFLk#lk429safnsf3", "2l23k23lSFLk#lk429saf\nsf3", ";2l23k23l SFLk#lk429'saf\nsf3 ", "abcDEF123@#$abc", "!@12asAS90123456fifty-one-charachers789012345678901","abCD$%789012345fifty-characters2345678901234567890"};

            runRegexTestCases(nameRegex                         , nameTestCases);
            runRegexTestCases(numberRegex                       , numberTestCases);
            runRegexTestCases(filenameRegex_InCurrentDir        , filenameTestCases);
            runRegexTestCases(filenameRegex_HasAcceptedExtension, filenameTestCases);
            runRegexTestCases(filenameRegex_HasRelativePath     , filenameTestCases);
            runRegexTestCases(passwordRegex                    , passwordTestCases);
	    }


        /**
         * 1) get first and last name from user
         */
        RegexValidator nameVerifiers[] = {new RegexValidator(nameRegex, true, "You must enter only letters")};
        firstName = getValidString("Enter your first name: ", STRING_INPUT_LENGTH, nameVerifiers);
        lastName  = getValidString("Enter your last name: " , STRING_INPUT_LENGTH, nameVerifiers);


        /**
         * 2) get 2 32 bit ints from user
         */
        RegexValidator numberVerifiers[] = {new RegexValidator(numberRegex, true, "You must only enter a maximum of ten digits, which may be preceded with an \"+\" or \"-\"")};
        long integerA = getValidInteger("Enter the first 32 bit integer: ", INT_INPUT_LENGTH, numberVerifiers);
        long integerB = getValidInteger("Enter the second 32 bit integer: ",INT_INPUT_LENGTH, numberVerifiers);
        long addResult  = integerA + integerB;
        long multResult = integerA * integerB;

        /**
         * 3) get input/output filename from user
         */
        RegexValidator filenameValidators[] = {
            new RegexValidator(filenameRegex_HasAcceptedExtension, true, "Only files with the extensions .txt and .text are allowed"),
            new RegexValidator(filenameRegex_HasRelativePath,      false, "You are not allowed to use relative paths"),
            new RegexValidator(filenameRegex_InCurrentDir,         true, "You may only specify files in the current directory"),
        };
        inputFile = getValidFile("Enter an input file path from the current directory"  , STRING_INPUT_LENGTH, filenameValidators, FileType.Input);
        outputFile = getValidFile("Enter an output file path from the current directory", STRING_INPUT_LENGTH, filenameValidators, FileType.Output);
        writeOutputFile(firstName, lastName, addResult, multResult, inputFile, outputFile);

        /**
         * 4) get password from user
         */
        RegexValidator passwordVerifiers[] = {new RegexValidator(passwordRegex, true, "Criteria not met.")};
	    doPasswordThing("Enter a password between 12 and 56 characters that contains one lower-case letter, one upper-case letter, one numerical digit, and one symbol.", passwordVerifiers);

    }// End run()

    private void doPasswordThing(String prompt, RegexValidator validators[]){
        try {

            boolean isValid = false;
            byte[] salt = new byte[8];
            // Generate salt
            SecureRandom random = SecureRandom.getInstance("SHA1PRNG");
            random.nextBytes(salt);

            // Get the hash of the password
            byte[] originalHash = SecretKeyFactory
                    .getInstance("PBKDF2WithHmacSHA1")
                    .generateSecret(new PBEKeySpec(getValidString(prompt, PWD_INPUT_LENGTH, validators).toCharArray(), salt, 65536, 128))
                    .getEncoded();


            while(!isValid){
                String passwordGuess = getValidString("Re-enter your password", PWD_INPUT_LENGTH, validators);

                // Get the hash of the guess
                byte[] guessHash = SecretKeyFactory
                        .getInstance("PBKDF2WithHmacSHA1")
                        .generateSecret(new PBEKeySpec(passwordGuess.toCharArray(), salt, 65536, 128))
                        .getEncoded();

                if(Arrays.equals(originalHash, guessHash)){
                    System.out.println("Congratulations, you matched your own password!");
                    isValid = true;
                }
                else{
                    System.out.println("- That was the wrong password");
                    isValid = false;
                }
            }// End while
        }// End try
        catch(Exception e){
            System.err.println();
            logError(e.getMessage());
            System.exit(1);
        }
    }// End doPasswordThing

    private String getValidString(String prompt, int inputLength, RegexValidator validators[]){
        if(prompt == null || validators == null)
            throw new NullPointerException();

        boolean isValid = false;
        UserInputResult thisInput = null;

        while(!isValid){
            isValid = true;

            System.out.println(prompt);
            System.out.println(">");
            thisInput = getInput(inputLength);

            if(thisInput.status == UserInputStatus.TooLarge){
                System.out.println("- You must enter fewer than " + (inputLength + 1) + " characters");
                isValid = false;
            }
            else if(thisInput.status == UserInputStatus.Empty){
                System.out.println("- You must enter something.");
            }
            else {
                for (RegexValidator thisValidator : validators){
                    if (!thisValidator.validate(thisInput.value)) {
                        isValid = false;
                        System.out.println("- " + thisValidator.failDescription);
                    }
                }
            }
        }// End while(!isValid)
        return thisInput.value;
    }// End getValidString

    private UserInputResult getInput(int length){
        StringBuilder value = new StringBuilder();

        InputStreamReader reader = new InputStreamReader(System.in);
        int copyCount = 0;
        char thisChar = 0;
        boolean exceededBuffer = false;

        try {
            while ((thisChar = (char)reader.read()) != '\n') {
                if(copyCount++ < length)
                    value.append(thisChar);
                else
                    exceededBuffer = true;
            }
        }
        catch(Exception e){
            logError(e.getMessage());
            System.out.println(e.getMessage());
            System.exit(1);
        }
        return new UserInputResult(value.toString(), (exceededBuffer) ? UserInputStatus.TooLarge : (copyCount == 0) ? UserInputStatus.Empty : UserInputStatus.WithinLimits);
    }// End getInput

    long getValidInteger(String prompt, int length, RegexValidator validators[]){
        boolean isValid = false;

        long returnValue = 0;

        while(!isValid) {
            String thisInput = getValidString(prompt, length, validators);
            isValid = true;
            try {
                returnValue = Long.parseLong(thisInput);
                if(returnValue > Integer.MAX_VALUE){
                    System.out.println("- The provided value exceeds the maximum value for a 32 bit signed integer");
                    isValid = false;
                }

                if(returnValue < Integer.MIN_VALUE){
                    System.out.println("- The provided value is less than the minimum value for a 32 bit signed integer");
                    isValid = false;
                }
            }
            catch(NumberFormatException e){ // This should never happen...
                logError(e.getMessage());
                System.out.println("- " + e.getMessage());
                isValid = false;
            }
        }// End while(!isValid)
        return returnValue;
    }// getValidInteger

    File getValidFile(String prompt, int length, RegexValidator validators[], FileType type){
        File validFile = null;

        boolean isValid = false;
        while(!isValid){
            String thisInput = getValidString(prompt, length,validators);

            try {
                Path thisPath = Paths.get(thisInput);
                boolean exists = Files.exists(thisPath);

                isValid = true;
                if (type == FileType.Input) {
                    if(!exists){
                        System.out.println("- You must specify an input file that exists.");
                        isValid = false;
                    }
                    else{
                        BasicFileAttributeView attributes = Files.getFileAttributeView(thisPath, BasicFileAttributeView.class, LinkOption.NOFOLLOW_LINKS);
                        FileOwnerAttributeView ownerAttributes = Files.getFileAttributeView(thisPath, FileOwnerAttributeView.class, LinkOption.NOFOLLOW_LINKS);


                        // Make sure we are dealing with a regular file
                        if(!attributes.readAttributes().isRegularFile()){
                            System.out.println("- You must specify a regular file (directories, character/block devices, named pipes, sockets, and symbolic links are not allowed)");
                            isValid = false;
                        }

                        // Make sure the current user is actually the owner of the input file
                        if(!ownerAttributes.getOwner().getName().equals(userName)){
                            System.out.println("- You must specify a file that you own");
                            isValid = false;
                        }

                        // Make sure the current user has read privileges for the specified file
                        validFile = thisPath.toFile();
                        if(!validFile.canRead()){
                            System.out.println("- You must specify a file that you have read privileges to");
                            isValid = false;
                        }
                    }// End else
                } else if (type == FileType.Output) {
                    if(exists){
                        System.out.println("- You must specify a file that doesn't exist already");
                        isValid = false;
                    }
                    else{
                        validFile = thisPath.toFile();
                    }
                }// End else if
            }
            catch(Exception e){
                logError(e.getMessage());
                System.err.println(e.getMessage());
                System.exit(1);
            }
        }// End while(!isValid)
        return validFile;

    }// End getValidFile

    private void writeOutputFile(String firstName, String lastName, long addResult, long multResult, File inputFile, File outputFile){
        String newLine = System.lineSeparator();
        String header = firstName + newLine + lastName + newLine + addResult + newLine + multResult + newLine;

        byte[] thisByte = new byte[1];

        try {
            PrintStream printStream = new PrintStream(outputFile);
            BufferedInputStream reader = new BufferedInputStream(new FileInputStream(inputFile));
            printStream.print(header);

            while(reader.read(thisByte) > 0){
                printStream.write(thisByte);
            }
            printStream.close();
            reader.close();
        }
        catch(Exception e){
            logError(e.getMessage());
            System.err.println(e.getMessage());
            System.exit(1);
        }
    }// End writeOutputFile

    private void logError(String error){
        try {
            PrintStream printStream = new PrintStream(logFile);
            printStream.println((new SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format(new Date())) + " - " +  error);
        }
        catch(Exception e){
            System.err.println(e.getMessage());
            System.exit(1);
        }
    }

    private void runRegexTestCases(String regex, String inputs[]){
        System.out.println("---- Test Cases for Regex: " + regex + "---------");
        System.out.println("  Is Match  |  Input  ");
        System.out.println("----------------------");

        RegexValidator thisValidator = new RegexValidator(regex, true, "");

        for(String thisInput: inputs){
            System.out.printf("%12s|%12s",(thisValidator.validate(thisInput) ? "YES" : "NO"), "\"" + thisInput + "\"");
            System.out.println();
        }
        System.out.println();
    }
}

File: Assignment_6/MichaelPeterson/C/Assignment_6/assignment6.c
/**
 * CSCD 437
 * Fall 2016
 * Assignment: Defend Your Code
 *
 * Date: November 4, 2016
 *
 * Team : The Wise Guys
 *    Matt Mcgee
 *    Heath Knickerbocker
 *    Michael Peterson
 *
 */
#include "assignment6.h"

#include <stdio.h>
#include <regex.h>
#include <stdlib.h>
#include <assert.h>
#include <unistd.h>
#include <errno.h>
#include <memory.h>
#include <limits.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <time.h>
#include "md5.h"

int main() {
	// First things first, make sure this isn't being ran as root
	if(!getuid()){
		fprintf(stderr, "\nThis program must not be run as the root user!\n");
		exit(1);
	}

#if __GNUC__ < 4
	fprintf(stderr, "\nThis is only designed to be compiled using GCC 4.0 or newer\n");
	exit(1);
#endif
#ifdef _WIN32
	fprintf(stderr, "\nThis program is only designed to run in Linux\n");
	exit(1);
#endif
#ifdef _WIN64
	fprintf(stderr, "\nThis program is only designed to run in Linux\n");
	exit(1);
#endif

	// input buffers
	char firstName[IN_BUFF_LENGTH], lastName[IN_BUFF_LENGTH];
	char inFilename[IN_BUFF_LENGTH], outFilename[IN_BUFF_LENGTH];
	char intA[NUMBER_BUFF_LENGTH], intB[NUMBER_BUFF_LENGTH];

	// file pointers
	FILE *inputFile = NULL, *outputFile = NULL, *logFile = NULL;

	// regexes
	char *nameRegex                          = "^[a-zA-Z]{1,}$";
	char *numberRegex                        = "^(()|\\+|\\-)[0-9]{1,10}$";
	char *filenameRegex_InCurrentDir         = "^(\\.\\/|[^\\/])[a-zA-Z0-9\\s]+(\\.[a-zA-Z]{1,4})$";
	char *filenameRegex_HasAcceptedExtension = "\\.(text|txt)$"; // White list of valid extensions, we need more ideas ...
	char *filenameRegex_HasRelativePath      = "\\.\\.";
	char *passwordRegex_hasLowerCase         = "[a-z]+";
	char *passwordRegex_hasUpperCase         = "[A-Z]+";
	char *passwordRegex_hasDigit             = "[0-9]+";
	char *passwordRegex_hasPunct             = "[`~!@#$%^&*()_+=;\\.]+";


#ifdef TEST
	// Regex test cases
	char * nameTestCases[]     = {"", "\0", "\n", "Jhon", " Jhon", "1ll124lk1 1 4l ", "bob", "a", "13", "bob ", "0", "..", "Jo hn", 0};
	char * numberTestCases[]   = {"", "\0", "\n", "0", " 0", "0 ", " 0 ", "-0", "+0", "123", "-123", "+123", "- 123", "123-", "+0123546789", "+01235467891", " ", "apple", "abc", "-12345678901", "--12345678", "-+34567890", "- 34567890", "123456789-","123456789 +", "123456789++", "-123456789-", "(123*456)", "0000000000", "-0000000000", "*1234567890", "1-23456789", "--", "--0", "0000000000000001", 0};
	char * filenameTestCases[] = {"", "\0", "\n", "file.dog", "file.Txt", "file.txt", "./file/../.txt", "/file.txt", "./file.txt", "../file.txt", "/../file.txt", "./ /", "\\/bin\\/file.txt", "*.txt", "file.*",  0};
	char * passwordTestCases[] = {"", "\0", "\n", "abcde", "Hello there.8", "Slg3k4k23j4Dkj4k23j4kn234jh", ";2l23k23l SFLk#lk429' saf\nsf3 ", "abcDEF123@#$abc", "!@12asAS90123456fifty-one-charachers789012345678901","abCD$%789012345fifty-characters2345678901234567890", 0};

	runRegexTestCases(nameRegex                         , nameTestCases);
	runRegexTestCases(numberRegex                       , numberTestCases);
	runRegexTestCases(filenameRegex_InCurrentDir        , filenameTestCases);
	runRegexTestCases(filenameRegex_HasAcceptedExtension, filenameTestCases);
	runRegexTestCases(filenameRegex_HasRelativePath     , filenameTestCases);
	runRegexTestCases(passwordRegex_hasLowerCase        , passwordTestCases);
	runRegexTestCases(passwordRegex_hasUpperCase        , passwordTestCases);
	runRegexTestCases(passwordRegex_hasDigit            , passwordTestCases);
	runRegexTestCases(passwordRegex_hasPunct            , passwordTestCases);

#endif

	regexValidator emptyValidator = {NULL, 0, NULL};

	// Try to open a log file for output
	errno = 0;
	if(!(logFile = fopen("error.log", "w"))){
		fprintf(stderr, "\n%s\n", strerror(errno));
		exit(1);
	}

	/**
	 * 1) get first and last name from user
	 */
	regexValidator nameValidators[] = {{nameRegex, 1, "You must enter only letters"}, emptyValidator};
	getValidatedString("Enter your first name: ", firstName, IN_BUFF_LENGTH, nameValidators);
	getValidatedString("Enter your last name: " , lastName,  IN_BUFF_LENGTH, nameValidators);

	/**
	 * 2) get 2 32 bit ints from user
	 */
	regexValidator numberValidators[] = {{numberRegex, 1, "you may enter up to ten digits which may optionally be precedded by \"+\" or \"-\""}, emptyValidator};
	long long integerA = getVerifiedInteger("Enter the first 32 bit integer: " ,intA, NUMBER_BUFF_LENGTH, numberValidators);
	long long integerB = getVerifiedInteger("Enter the second 32 bit integer: ",intB, NUMBER_BUFF_LENGTH, numberValidators);
	long long addResult  = integerA + integerB;
	long long multResult = integerA * integerB;

	/**
	 * 3) get input/output filename from user
	 */
	regexValidator filenameValidators[] = {
			{filenameRegex_HasAcceptedExtension, 1, "only files with the extensions .txt and .text are allowed"},
			{filenameRegex_HasRelativePath,      0, "you are not allowed to use relative paths"},
			{filenameRegex_InCurrentDir,         1, "you may only specify files in the current directory"},
			emptyValidator
	};
	inputFile  = getValidFile("Enter an input file path from the current directory" , inFilename , IN_BUFF_LENGTH, filenameValidators, INPUT_FILE);
	outputFile = getValidFile("Enter an output file path from the current directory", outFilename, IN_BUFF_LENGTH, filenameValidators, OUTPUT_FILE);
	writeOutputFile(firstName, lastName, addResult, multResult, inputFile, outputFile, logFile);

	/**
	 * 4) get password from user
	 */
	regexValidator passwordValidators[] = {
			{passwordRegex_hasLowerCase, 1, "You must enter at least one lower case letter."},
			{passwordRegex_hasUpperCase, 1, "You must enter at least one upper case letter"},
			{passwordRegex_hasDigit,     1, "You must enter a digit"},
			{passwordRegex_hasPunct,     1, "You must enter at least on punctuation character"},
			emptyValidator
	};

	doPasswordThing("Enter a password of 12-56 characters, with at least one capital, one lower case, one digit, and one special ( `~!@#$%^&*()_+= ) symbol: ", passwordValidators);

	fclose(inputFile);
	fclose(outputFile);
	fclose(logFile);

	return 0;
}// End Main


/**************************************************************
*                                                            *
*                      HELPER FUNCTIONS                      *
*                                                            *
*      |        |        |        |        |        |        *
*      v        v        v        v        v        v        *
**************************************************************/


int runRegexTestCases(char *regex, char *testCases[]) {
#ifdef TEST
	printf("\n\n---- Test Cases for Regex: \"%s\" --------\n", regex);
	printf("  Is Match  |  Input  \n");
	printf("----------------------\n");
#endif
	char *thisTestCase;
	int i = 0, passCount = 0, havePass = 0;
	while ((thisTestCase = testCases[i++]) != 0) {
		passCount += (havePass = isRegexMatch(regex, thisTestCase)) ? 1 : 0;
#ifdef TEST
		printf("%8s    |    \"%s\"\n", (havePass) ? "YES" : "No", thisTestCase);
#endif
	}
	return passCount;
}

int isRegexMatch(char *regex, char *input) {
	if (regex == NULL || input == NULL)
		return 0;
	char errorMessage[100];

	regex_t regext;
	int reti = regcomp(&regext, regex, REG_EXTENDED);
	if (reti) {
		regerror(reti, &regext, errorMessage, sizeof(errorMessage));
		fprintf(stderr, "%s\n", errorMessage);
		exit(1);
	}

	reti = regexec(&regext, input, 0, NULL, 0);
	regfree(&regext);

	if (!reti) {
		return 1;
	} else if (reti == REG_NOMATCH) {
		return 0;
	} else {
		regerror(reti, &regext, errorMessage, sizeof(errorMessage));
		fprintf(stderr, "%s\n", errorMessage);
		return 0;
	}
}

void getValidatedString(char *prompt, char *inputBuffer, int inputBufferSize, regexValidator validators[]) {
	assert (prompt != NULL && inputBuffer != NULL && inputBufferSize != 0 && validators != NULL);

	int isValid = 0;

	while (!isValid) {
		printf("\n%s\n>", prompt);
		int readResult = readInput(inputBuffer, inputBufferSize);
		if (readResult == READ_EXCEEDED_BUFFER) {
			printf("\nYou must enter %d or fewer characters. Try again:\n", inputBufferSize - 1 );
			isValid = 0;
		} else if (readResult == READ_ZERO_BYTES) {
			printf("\nYou must enter something...\n");
			isValid = 0;
		} else {
			int validatorIndex = 0;
			int failCount = 0;
			regexValidator thisValidator;
			while ((thisValidator = validators[validatorIndex++]).regex != NULL) {
				if (thisValidator.shouldMatch != isRegexMatch(thisValidator.regex, inputBuffer)) {
					failCount++;
					isValid = 0;
					printf("\n- %s\n", thisValidator.failDescription);
				}
			}
			if (failCount == 0)
				isValid = 1;
		}
	}
}

int readInput(char *inputBuffer, int bufferLength) {
	unsigned char thisChar;
	int exceededBuffer = 0;
	int copyCount = 0;
	while ((thisChar = getchar()) != '\n') {
		if (copyCount < bufferLength - 1)
			inputBuffer[copyCount++] = thisChar;
		else
			exceededBuffer = 1;
	}
	inputBuffer[copyCount] = '\0';
	return (exceededBuffer) ? READ_EXCEEDED_BUFFER : (copyCount == 0) ? READ_ZERO_BYTES : READ_WITHIN_BUFFER;
}

long long getVerifiedInteger(char *prompt, char *inputBuffer, int inputBufferSize, regexValidator validators[]){
	int isValid = 0;
	long long returnValue = 0;
	while(!isValid){
		getValidatedString(prompt, inputBuffer, inputBufferSize, validators);
		errno = 0;
		returnValue = strtoll(inputBuffer, NULL, 10);
		isValid = 1;
		if(returnValue > INT_MAX){
			printf("\n- The provided value exceeds the maximum value for a 32 bit signed integer");
			isValid = 0;
		}

		if(returnValue < INT_MIN){
			printf("\n- The provided value is less then the minimum value for a 32 but signed integer");
			isValid = 0;
		}

		if(errno){ // This should never happen...
			printf("%s\n", strerror(errno));
			isValid = 0;
		}
	}
 return returnValue;
}

FILE* getValidFile(char *prompt, char *inputBuffer, int inputBufferSize, regexValidator validators[], int fileType){
	assert(fileType == INPUT_FILE || fileType == OUTPUT_FILE);

	FILE* validFile = NULL;

	int isValid = 0;
	while(!isValid){
		getValidatedString(prompt, inputBuffer, inputBufferSize, validators);

		errno = 0;
		struct stat thisStat;
		int statReturnValue = lstat(inputBuffer, &thisStat);

		isValid = 1;
		if(fileType == INPUT_FILE){
			// Fail if something went wrong trying to get the input files status (this includes files that don't exist)
			if (statReturnValue) {
				printf("\n- %s\n", strerror(errno));
				isValid = 0;
			}
			else{
				mode_t mode = thisStat.st_mode;
				int ownerId = thisStat.st_uid;

				// Make sure we are dealing with a regular file
				if(!S_ISREG(mode)){
					printf("\n- You must specify a regular file (directories, character/block devices, named pipes, sockets, and symbolic links are not allowed)\n");
					isValid = 0;
				}

				// Make sure the current user is actually the owner of the input file
				if(ownerId != (int)getuid()){
					printf("\n- You must specify a file that you own\n");
					isValid = 0;
				}

				// Make sure the owner has read privileges
				if(!(mode & S_IRUSR)){
					printf("\n- You must specify a file that you have read privileges to\n");
					isValid = 0;
				}

				errno = 0;
				// Try opening the file, complain if something goes wrong
				if(isValid && !(validFile = fopen(inputBuffer, "r"))){
					printf("\n- %s\n", strerror(errno));
					isValid = 0;
				}
			}// End else
		}// End if(fileType == INPUT_FILE)
		else if(fileType == OUTPUT_FILE){
			if(errno != ENOENT){
				printf("\n- You must specify a file that doesn't exist yet\n");
				isValid = 0;
			}
			else{

				errno = 0;
				// Try opening a new file for output, if something goes wrong complain
				if(!(validFile = fopen(inputBuffer, "w"))){
					printf("\n- %s\n", strerror(errno));
					isValid = 0;
				}
			}// End else
		}// End if(fileType == OUTPUT_FILE)
	}// End while(!isValid)

	return validFile;
}

void writeOutputFile(char* firstName, char* lastName, long long addResult, long long multResult, FILE* inputFile, FILE* outputFile, FILE* logFile){
	char multResultString[21];
	char addResultString[12];
	char outputFileHeader[IN_BUFF_LENGTH*2 + sizeof(multResultString) + sizeof(addResultString) + 4];

	snprintf(multResultString, sizeof(multResultString), "%lld", multResult);
	snprintf(addResultString , sizeof(addResultString) , "%lld", addResult);
	snprintf(outputFileHeader, sizeof(outputFileHeader), "%s\n%s\n%s\n%s\n", firstName, lastName, addResultString, multResultString);

	char thisChar;
	int headerIndex = 0;
	errno = 0;
	while((thisChar = outputFileHeader[headerIndex++]) != '\0') {
		putc(thisChar, outputFile);
		if (errno) {
			fprintf(stderr, "\n%s\n", strerror(errno));
			fclose(inputFile);
			fclose(outputFile);
			exit(1);
		}
	}

	while(!feof(inputFile)){
		errno = 0;
		int thisChar = fgetc(inputFile);
		if(feof(inputFile))
			break;
		if(errno){
			char* errorOutput = strerror(errno);
			printf("\n%s\n",errorOutput);
			fprintf(logFile, "%s\n",errorOutput);
			fclose(outputFile);
			fclose(inputFile);
			exit(1);
		}
		else{
			errno = 0;
			putc(thisChar, outputFile);
			if(errno){
				char* errorOutput = strerror(errno);
				printf("\n%s\n",errorOutput);
				fprintf(logFile, "%s\n",errorOutput);
				fclose(inputFile);
				fclose(outputFile);
				exit(1);
			}
		}
	}

}

void doPasswordThing(char* prompt, regexValidator validators[]){
	char password[64];
	char passwordGuess[sizeof(password)];
	unsigned int originalHash[4];
	unsigned int guessHash[4];
	char salt[8];

	// generate a random salt
	srand(clock());
	int i;
	for(i = 0; i < (int)sizeof(salt); ++i)
		salt[i] = (char)((rand() % (UCHAR_MAX - 1)) +1);

	// get the password from the user
	getValidatedString(prompt,password, sizeof(password) - 8, validators);
	int passwordLength = strlen((char*)password);

	// Append the salt to the end of the string and fill the rest with '\0's
	int saltIndex = 0;
	for(i = passwordLength; i < passwordLength + (int)sizeof(salt); ++i)
		password[i] = salt[saltIndex++];
	for(; i < (int)sizeof(password); ++i)
		password[i] = '\0';

	// Generate the md5 hash
	md5_vfy((unsigned char*)password, sizeof(password), originalHash, originalHash+1, originalHash+2, originalHash+3);

	// Empty the original password
	for(i = 0; i < (int)sizeof(password); ++i)
		password[i] = '\0';

	int foundPassword = 0;
	while(!foundPassword){
		foundPassword = 1;

		// Get the password guess from the user
		getValidatedString("Enter your password guess", passwordGuess, sizeof(passwordGuess), validators);
		int passwordGuestLength = strlen((char*)passwordGuess);

		// Append the salt to the end of the string and fill the rest with '\0's
		saltIndex = 0;
		for(i = passwordLength; i < passwordGuestLength + (int)sizeof(salt); ++i)
			passwordGuess[i] = salt[saltIndex++];
		for(; i < (int)sizeof(password); ++i)
			passwordGuess[i] = '\0';

		// Hash the password
		md5_vfy((unsigned char*)passwordGuess, sizeof(password), guessHash, guessHash+1, guessHash+2, guessHash+3);

		for(i = 0; i < 4; ++i)
			if(originalHash[i] != guessHash[i])
				foundPassword = 0;

		if(foundPassword){
			printf("\ncongratulations, you found the password!!\n\n");
			break;
		}
		else
			printf("\n- That was the wrong password\n");
	}

}
